stdlib/fun.mli

```diff
@@ -31,3 +33,30 @@ val flip : ('a -> 'b -> 'c) -> ('b -> 'a -> 'c)
 val negate : ('a -> bool) -> ('a -> bool)
 (** [negate p] is the negation of the predicate function [p]. For any
     argument [x], [(negate p) x] is [not (p x)]. *)
+
+(** {1 Exception handling} *)
+
+val protect : finally:(unit -> unit) -> (unit -> 'a) -> 'a
+(** [protect ~finally work] invokes [work ()] and then [finally ()]
+    before [work ()] returns with its value or an exception. In the
+    latter case the exception is re-raised after [finally ()]. If
+    [finally ()] raises an exception, then the exception
+    {!Finally_raised} is raised instead.
+
+    [protect] can be used to enforce local invariants whether the
+    function returns normally or raises an exception. However, it does
+    not protect against unexpected exceptions raised inside [finally
+    ()] such as {!Stdlib.Out_of_memory}, {!Stdlib.Stack_overflow}, or
+    asynchronous exceptions raised by signal handlers
+    (e.g. {!Sys.Break}). *)
+
+exception Finally_raised of exn
+(** Raised by [protect ~finally work] when [finally] raises an
+    exception. The argument is the exception that was raised.
+
+    [Finally_raised] can contain an unexpected exception raised in
+    [finally] and/or shadow one raised in [work], such as
```

Comment:
Sorry to insist on this but this is confusing `Finally_raised` never contains an exception raised in `work`.

Reply:
Let's make this simple this is the exact doc string I would like to see. I don't think more needs to be said than this:

```
(** [Finally_raised exn] is raised by [protect ~finally work] when [finally] raises an 
    unexpected exception [exn]. Note that if this happens and [work] returned by raising
    an exception the latter exception is lost. *)
```

Reply:
I have changed it to just that.